File: python/problems/leetcode/0406_queue_reconstruction_by_height/solution.py
"""LeetCode 406: Queue Reconstruction by Height."""

from __future__ import annotations


class Solution:
    def reconstructQueue(self, people: list[list[int]]) -> list[list[int]]:
        people.sort(key=lambda p: (-p[0], p[1]))
        res: list[list[int]] = []
        for p in people:
            res.insert(p[1], p)
        return res


if __name__ == "__main__":
    out = Solution().reconstructQueue(
        [[7, 0], [4, 4], [7, 1], [5, 0], [6, 1], [5, 2]]
    )
    assert out == [[5, 0], [7, 0], [5, 2], [6, 1], [4, 4], [7, 1]]
    print("0406 OK")

File: cpp/problems/leetcode/0406_queue_reconstruction_by_height/solution.cpp
// LeetCode 406
#include <alg_std.hpp>
#include <cassert>
using namespace std;

class Solution {
public:
    vector<vector<int>> reconstructQueue(vector<vector<int>>& people) {
        sort(people.begin(), people.end(),
             [](const vector<int>& a, const vector<int>& b) {
                 return a[0] != b[0] ? a[0] > b[0] : a[1] < b[1];
             });
        vector<vector<int>> res;
        for (auto& p : people) res.insert(res.begin() + p[1], p);
        return res;
    }
};

int main() {
    vector<vector<int>> in{{7, 0}, {4, 4}, {7, 1}, {5, 0}, {6, 1}, {5, 2}};
    auto out = Solution().reconstructQueue(in);
    vector<vector<int>> expect{{5, 0}, {7, 0}, {5, 2}, {6, 1}, {4, 4}, {7, 1}};
    assert(out == expect);
    cout << "0406 OK" << endl;
    return 0;
}
